src/boot/efi/devicetree.c

```diff
@@ -0,0 +1,164 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <efi.h>
+
+#include "devicetree.h"
+#include "missing_efi.h"
+#include "util.h"
+
+#define FDT_V1_SIZE (7*4)
+
+#ifndef EFI_DT_FIXUP_PROTOCOL_GUID
+#define EFI_DT_FIXUP_PROTOCOL_GUID \
+        { 0xe617d64c, 0xfe08, 0x46da, {0xf4, 0xdc, 0xbb, 0xd5, 0x87, 0x0c, 0x73, 0x00} }
+#endif
+
+#define EFI_DT_FIXUP_PROTOCOL_REVISION 0x00010000
+
+/* Add nodes and update properties */
+#define EFI_DT_APPLY_FIXUPS    0x00000001
+/*
+ * Reserve memory according to the /reserved-memory node
+ * and the memory reservation block
+ */
+#define EFI_DT_RESERVE_MEMORY  0x00000002
+
+typedef struct _EFI_DT_FIXUP_PROTOCOL EFI_DT_FIXUP_PROTOCOL;
+
+typedef EFI_STATUS (EFIAPI *EFI_DT_FIXUP) (
+        IN EFI_DT_FIXUP_PROTOCOL *This,
+        IN VOID                  *Fdt,
+        IN OUT UINTN             *BufferSize,
+        IN UINT32                Flags);
+
+struct _EFI_DT_FIXUP_PROTOCOL {
+        UINT64         Revision;
+        EFI_DT_FIXUP   Fixup;
+};
+
+static EFI_STATUS devicetree_allocate(struct devicetree_state *state, UINTN size) {
+        UINTN pages = DIV_ROUND_UP(size, EFI_PAGE_SIZE);
+        EFI_STATUS err;
+
+        assert(state);
+
+        err = uefi_call_wrapper(BS->AllocatePages, 4, AllocateAnyPages,
+                        EfiACPIReclaimMemory, pages, &state->addr);
+        if (EFI_ERROR(err))
+                return err;
+
+        state->pages = pages;
+        return err;
+}
+
+static UINTN devicetree_allocated(struct devicetree_state *state) {
+        assert(state);
+        return state->pages * EFI_PAGE_SIZE;
+}
+
+static VOID *devicetree_ptr(struct devicetree_state *state) {
+        assert(state);
+        assert(state->addr <= UINTPTR_MAX);
+        return (VOID *)(uintptr_t)state->addr;
+}
+
+static EFI_STATUS devicetree_fixup(struct devicetree_state *state, UINTN len) {
+        EFI_DT_FIXUP_PROTOCOL *fixup;
+        UINTN size;
+        EFI_STATUS err;
+
+        assert(state);
+
+        err = LibLocateProtocol(&(EFI_GUID)EFI_DT_FIXUP_PROTOCOL_GUID, (VOID **)&fixup);
+        if (EFI_ERROR(err)) {
+                Print(L"Could not locate device tree fixup protocol, skipping.\n");
```

Comment:
I dunno, I think this should just be silent case...

Reply:
In the future I'd agree. But for now the dt fixup protocol seems to only be implemented in u-boot, and I think it'd be nice for anyone trying out other EFI implementations to know why things might not work as expected.

Reply:
Then you should be using log_error_stall(), cause otherwise that message is gonna vanish quickly.